src/model_repository_manager.h

```diff
@@ -260,65 +267,119 @@ class ModelRepositoryManager {
-      const std::set<std::string>& modified,
-      std::set<std::string>* deleted_dependents = nullptr);
+      const std::set<std::string>& modified, const ModelInfoMap& model_infos,
+      DependencyGraph* updated_graph,
+      std::set<std::string>* deleted_dependents = nullptr) const;
+
+  /// Helper function to deep copy the dependency graph
+  /// \param new_graph The new graph to be cleared and copied into
+  void CopyDependencyGraph(DependencyGraph* new_graph) const;
+
+  /// Helper function to re-map dependency graph nodes pointers from old nodes
+  /// to new nodes.
+  /// The function search for matching pointers in the new_nodes first, if not
+  /// found, then the matching pointer must be in the ref_nodes.
+  /// \param ref_nodes The reference nodes for pointers not in new_nodes
+  /// \param new_nodes The new nodes to be re-mapped
+  void ReMapDependencyGraphPointers(
+      const std::unordered_map<std::string, std::unique_ptr<DependencyNode>>*
+          ref_nodes,
+      std::unordered_map<std::string, std::unique_ptr<DependencyNode>>*
+          new_nodes) const;
 
   /// Helper function to uncheck the nodes because the model that they depends
   /// on has changed. The unchecked nodes will be validated again.
   /// The function will be call recursively to uncheck all downstreams.
   /// \param downstreams The nodes to be unchecked.
   /// \param updated_nodes Return the nodes that have been unchecked
-  void UncheckDownstream(NodeSet* downstreams, NodeSet* updated_nodes);
+  void UncheckDownstream(NodeSet* downstreams, NodeSet* updated_nodes) const;
 
   /// Helper function to construct the edges between nodes in dependency graph.
+  /// \param graph The dependency graph.
   /// \param updated_node The node that is newly added or modified.
   /// \return True if the node represents an ensemble model. False otherwise.
-  bool ConnectDependencyGraph(DependencyNode* updated_node);
+  bool ConnectDependencyGraph(
+      DependencyGraph* graph, DependencyNode* updated_node) const;
 
   /// Get the model info for a named model.
+  /// \param model_infos Map of model names to its infomation.
   /// \param name The model name.
   /// \param model_info Returns the model information.
   /// \return OK if found, NOT_FOUND otherwise.
-  Status GetModelInfo(const std::string& name, ModelInfo** model_info);
+  Status GetModelInfo(
+      const ModelInfoMap& model_infos, const std::string& name,
+      ModelInfo** model_info) const;
+
+  /// Helper function to deep copy the model infos
+  /// \param new_infos The new model infos to be cleared and copied into
+  void CopyModelInfos(ModelInfoMap* new_infos) const;
+
+  /// Check if a dependency node (model) is currently loading/unloading
+  /// \param dependency_node The dependency node to check
+  /// \return OK if NOT in transit, INVALID_ARG if in transit
+  Status InTransit(const DependencyNode* dependency_node) const;
```

Comment:
The function is to be tapped into parts of the dependency graph operations to check if a given node is in transition, and be wrapped in `RETURN_IF_ERROR`. If a node is transiting, the request to load/unload model will return/fail with `INVALID_ARG` and the conflict explanation. Otherwise, the code continues. This is intended to return a `Status` and be chained backward throughout the returns if errors, to act as a shortcut to stop the execution.

Reply:
Yes. Updated.